### PR TITLE
state that casting away const and then mutating is undefined behaviour

### PR DESCRIPTION
**Pull request by aG0aep6G**
Related discussion: http://forum.dlang.org/post/riiehqozpkyluhhifwha@forum.dlang.org

I think there's a consensus now that casting away const and then mutating is always undefined behaviour.

Drive-by style fixes: title case, "immutableness" -> "immutability"

### CODE REVIEW

**Comment by Hackerpilot** [comment]
:white_check_mark:

**Comment by MartinNowak** [comment]
> Note that casting away a const qualifier and then mutating is undefined behaviour, too, even when the referenced data is mutable.

This sounds too harsh to me, I'd like to have a word on this from @WalterBright.

**Comment by jmdavis** [comment]
> This sounds too harsh to me, I'd like to have a word on this from @WalterBright.

@MartinNowak I totally agree that it would be a good idea to get Walter's approval on this, but I don't see how it's too harsh. It's been brought up time and time again that D's const is physical const and that any attempt to mutate it is undefined behavior. Allowing anything else makes it so that you cannot rely on a const object not being mutated via a const reference. Walter has made it abundantly clear in the past that D's const is actually supposed to provide the guarantee that an object cannot be mutated via a const reference and that the compiler is free to assume that fact. Allowing the casting away of const and then mutating to _ever_ be defined behavior violates this and makes it so that we're back to C++'s const except that it's transitive, and you have to be careful about the possibility of the object actually being immutable. But it's still basically C++'s const and not physical const at all if casting away and mutating is defined behavior. No D program should _ever_ be casting away const to mutate. If it's doing that, it should not be using const.

**Comment by aG0aep6G** [comment]
Ping @WalterBright? (Does he get notices about these @ mentions?)

Btw, I just came across this on <http://dlang.org/cpp_interface.html>:
```
// No:
// ptr to const int
const(int)* p;
int* q = cast(int*)p;
*q = 3;   // undefined behavior
```

File: const3.dd [hunk 1]
```diff
@@ -309,9 +309,26 @@ int* q = cast(int*)p;
...
 	Casting is, as always, a blunt and effective instrument, and
 	when using it to cast away immutable-correctness, one must assume
-	the responsibility to ensure the immutableness of the data, as
+	the responsibility to ensure the immutability of the data, as
 	the compiler will no longer be able to statically do so.
 	)
+
+	$(P
+	Note that casting away a const qualifier and then mutating is undefined
+	behaviour, too, even when the referenced data is mutable. This is so that
```
Comment (WalterBright) [inline const3.dd]: Please use American spelling, i.e. `behavior`

**Comment by WalterBright** [comment]
LGTM after spell correction.

**Comment by aG0aep6G** [comment]
> Please use American spelling, i.e. behavior

fixed

**Comment by aG0aep6G** [comment]
Ping @MartinNowak @jmdavis. Walter gave his ok. Anything else to do here?

**Pull request merged**